### PR TITLE
Grid locator

### PR DESCRIPTION
**Pull request by Jacob-Stevens-Haas**
Substantial refactoring going on here.   Big changes are (a) moving gridsearch to its own package to limit the coupling between it and the odes/pdes experiments, and (b) introducing the `gen_experiments.gridsearch.GridLocator` object, which allows one to specify criteria to match intermediate points in the gridsearch.

To test it out, run any existing experiment, build a `GridLocator` object, and find the matching points with `gridsearch.find_gridpoints(your_locator, results["plot_data"], results)`

### CODE REVIEW

File: src/gen_experiments/config.py [hunk 1]
```diff
@@ -41,77 +49,43 @@ def addn(x):
...
-        (
-            {"sim_params.noise_abs": 0.1},
-            {"sim_params.noise_abs": 0.5},
-            {"sim_params.noise_abs": 1},
-            {"sim_params.noise_abs": 2},
-            {"sim_params.noise_abs": 4},
-            {"sim_params.noise_abs": 8},
+        GridLocator(
+            ...,
+            (..., ...),
```
Comment (yb6599) [inline src/gen_experiments/config.py:61]: What does `...` signify here in the GridLocator arguments? Is it meant to be taking all the metrics or is it empty?

File: src/gen_experiments/config.py [hunk 2]
```diff
@@ -41,77 +49,43 @@ def addn(x):
...
+                {"sim_params.noise_abs": 0.5},
+                {"sim_params.noise_abs": 1},
+                {"sim_params.noise_abs": 2},
+                {"sim_params.noise_abs": 4},
+                {"sim_params.noise_abs": 8},
+            ),
         ),
     ),
-    "test-absrel3": _PlotPrefs(
+    "absrel-newloc": _PlotPrefs(
```
Comment (yb6599) [inline src/gen_experiments/config.py:72]: For running the experiments, which one of the `plot_prefs` do I use? `test_absrel12` or `absrel-newloc`?

**Review by yb6599 — COMMENTED** [review]
LGTM

File: src/gen_experiments/config.py (hunk 1 above)
Comment (Jacob-Stevens-Haas) [reply src/gen_experiments/config.py:61]: Nice!  Yeah, I wanted to borrow some notation from numpy that means "all elements".  They're both what are called "singletons".  Only one `NoneType` object exists in a python program, just like only one `EllipsisType` object ever exists.

I suppose I could have used `None` instead.  That may be an arguably better way, I just wanted to use an object that people associate with "all" of something.

File: src/gen_experiments/config.py (hunk 2 above)
Comment (Jacob-Stevens-Haas) [reply src/gen_experiments/config.py:72]: `absrel-newloc`, but double check me that it's meaning is clear, given the `GridLocator` argument

**Pull request merged**